File: ChessEngine/illegal_removal.py
from offsets import *
from pawn import check_pawn
from parsing import parse_move
from knight import check_knight
from rook import check_rook
from bishop import check_bishop
from queen import check_queen
from king import check_king, find_king_pos, find_specific_king_pos
from move import make_move

def remove_illegal_moves(current_board: dict, moves_list: dict, can_passant: list, castles: list, player: str) -> dict:
    """
    This Python function removes illegal moves for a player based on the current board state and
    available moves.
    
    :param current_board: The `current_board` parameter is a dictionary representing the current state
    of the chess board. It likely contains information about the positions of different pieces on the
    board
    :type current_board: dict
    :param moves_list: The `moves_list` parameter seems to be a dictionary containing moves for each
    player. The keys 'w' and 'b' likely represent the white and black players, respectively. The values
    are lists of moves that each player can make
    :type moves_list: dict
    :param player: The `player` parameter in the `remove_illegal_moves` function represents the current
    player whose moves are being checked for legality. It can have a value of either 'w' for white
    player or 'b' for black player
    :type player: str
    :return: The function `remove_illegal_moves` is returning a modified `moves_list_copy` dictionary
    with illegal moves removed for the specified player.
    """
    current_board2 = current_board.copy()
    moves_list_copy = moves_list.copy()
    moves_list2 = {'w':[], 'b':[]}
    king_pos = find_king_pos(current_board2, player)
   
    opponent = 'w' if player == 'b' else 'b'
    deleted_moves = []
    for mov in moves_list_copy[player]:
        if mov in castles:
            continue
        current_board2 = current_board.copy()
        moves_list2 = {'w':[], 'b':[]}
        
        make_move(mov[0], mov[1], current_board2, can_passant, castles, moves_list, player, check_promotion = False)
        check_bishop(current_board2, moves_list2)
        check_knight(current_board2, moves_list2)
        check_king(current_board2, moves_list2)
        check_queen(current_board2, moves_list2)
        check_rook(current_board2, moves_list2)
        check_pawn(current_board2, moves_list2)
        
        for mov2 in moves_list2[opponent]:
            
            if mov2[1] == find_specific_king_pos(current_board2, player):
                deleted_moves.append(mov)
    
    for mov in moves_list[player]:
        if player == 'w' and current_board[mov[1]].islower():
            deleted_moves.append(mov)
        if player == 'b' and current_board[mov[1]].isupper():
            deleted_moves.append(mov)
        
    for mov in deleted_moves:
        if mov in moves_list_copy[player]:
            moves_list_copy[player].remove(mov)
    
                
    opponent = 'w' if player == 'b' else 'b'
    
    moves_list_copy2 = {'w':[],'b':[]}
    check_bishop(current_board, moves_list_copy2)
    check_knight(current_board, moves_list_copy2)
    check_pawn(current_board, moves_list_copy2)
    check_rook(current_board, moves_list_copy2)
    check_queen(current_board, moves_list_copy2)
    check_king(current_board, moves_list_copy2)   
    for mov in moves_list_copy[player]:
        if mov not in moves_list_copy2[player] and not (mov[0] in king_pos or mov[1] in king_pos):
            moves_list_copy[player].remove(mov)
    
    return moves_list_copy

File: ChessEngine/chessEngineCPP/illegal_removal.cpp
#include <iostream>
#include <map>
#include <vector>
#include <algorithm>
#include <cctype>
#include "pawn.h"
#include "knight.h"
#include "rook.h"
#include "bishop.h"
#include "queen.h"
#include "king.h"
#include "move.h"

using namespace std;

// Function to remove illegal moves from the move list
map<char, vector<pair<pair<int, int>, pair<int, int>>>> remove_illegal_moves(
    map<pair<int, int>, char>& current_board, 
    map<char, vector<pair<pair<int, int>, pair<int, int>>>>& moves_list, 
    vector<pair<pair<int, int>, pair<int, int>>>& can_passant, 
    vector<pair<pair<int, int>, pair<int, int>>>& castles, 
    char player) 
{
    // Copying current board and move list
    map<pair<int, int>, char> current_board2 = current_board;
    map<char, vector<pair<pair<int, int>, pair<int, int>>>> moves_list_copy = moves_list;
    map<char, vector<pair<pair<int, int>, pair<int, int>>>> moves_list2 = { {'w', {}}, {'b', {}} };

    // Find the king position of the current player
    pair<int, int> king_pos = find_king_pos(current_board2, player).front();
    char opponent = (player == 'b') ? 'w' : 'b';
    vector<pair<pair<int, int>, pair<int, int>>> deleted_moves;

    // Iterating through the move list for the current player
    for (const auto& mov : moves_list_copy[player]) {
        // Skip castling moves
        if (find(castles.begin(), castles.end(), mov) != castles.end()) continue;

        // Copy the board state again and reset moves_list2
        current_board2 = current_board;
        moves_list2 = { {'w', {}}, {'b', {}} };

        // Make the move and check the board state for the opponent's potential responses
        make_move(mov.first, mov.second, current_board2, can_passant, castles, moves_list, player, false);

        check_bishop(current_board2, moves_list2);
        check_knight(current_board2, moves_list2);
        check_king(current_board2, moves_list2);
        check_queen(current_board2, moves_list2);
        check_rook(current_board2, moves_list2);
        check_pawn(current_board2, moves_list2);

        // Check if the opponent can attack the king after this move
        for (const auto& mov2 : moves_list2[opponent]) {
            if (mov2.second == find_specific_king_pos(current_board2, player)) {
                deleted_moves.push_back(mov);
            }
        }
    }

    // Remove moves that capture opponent's piece incorrectly
    for (const auto& mov : moves_list[player]) {
        if (player == 'w' && current_board.count(mov.second) && islower(current_board[mov.second])) {
            deleted_moves.push_back(mov);
        }
        if (player == 'b' && current_board.count(mov.second) && isupper(current_board[mov.second])) {
            deleted_moves.push_back(mov);
        }
    }

    // Remove deleted moves from the original move list
    for (const auto& mov : deleted_moves) {
        moves_list_copy[player].erase(remove(moves_list_copy[player].begin(), moves_list_copy[player].end(), mov), moves_list_copy[player].end());
    }

    // Rechecking all moves after removal to ensure consistency
    map<char, vector<pair<pair<int, int>, pair<int, int>>>> moves_list_copy2 = { {'w', {}}, {'b', {}} };
    check_bishop(current_board, moves_list_copy2);
    check_knight(current_board, moves_list_copy2);
    check_pawn(current_board, moves_list_copy2);
    check_rook(current_board, moves_list_copy2);
    check_queen(current_board, moves_list_copy2);
    check_king(current_board, moves_list_copy2);

    for (auto it = moves_list_copy[player].begin(); it != moves_list_copy[player].end(); ) {
        if (find(moves_list_copy2[player].begin(), moves_list_copy2[player].end(), *it) == moves_list_copy2[player].end() &&
            it->first != king_pos && it->second != king_pos) {
            it = moves_list_copy[player].erase(it);
        } else {
            ++it;
        }
    }

    // Remove invalid king moves (if any)
    for (auto it = moves_list_copy[player].begin(); it != moves_list_copy[player].end(); ) {
        if (tolower(current_board[it->first]) == 'k' && 
            abs(it->first.first - it->second.first) + abs(it->first.second - it->second.second) > 2) {
            it = moves_list_copy[player].erase(it);
        } else {
            ++it;
        }
    }

    return moves_list_copy;
}
